presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -126,11 +144,147 @@ public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
+        }
+    }
+
+    private static class PushDownNegationsExpressionRewriter
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteNotExpression(NotExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getValue() instanceof LogicalBinaryExpression) {
+                LogicalBinaryExpression child = (LogicalBinaryExpression) node.getValue();
+                List<Expression> predicates = extractPredicates(child);
+                List<Expression> negatedPredicates = predicates.stream()
+                        .map(predicate -> treeRewriter.rewrite((Expression) new NotExpression(predicate), context))
+                        .collect(toList());
+                return combinePredicates(child.getType().flip(), negatedPredicates);
+            }
+            else if (node.getValue() instanceof ComparisonExpression && ((ComparisonExpression) node.getValue()).getType() != IS_DISTINCT_FROM) {
+                ComparisonExpression child = (ComparisonExpression) node.getValue();
+                return new ComparisonExpression(
+                        child.getType().negate(),
+                        treeRewriter.rewrite(child.getLeft(), context),
+                        treeRewriter.rewrite(child.getRight(), context));
+            }
+            else if (node.getValue() instanceof NotExpression) {
+                NotExpression child = (NotExpression) node.getValue();
+                return treeRewriter.rewrite(child.getValue(), context);
+            }
+
+            return new NotExpression(treeRewriter.rewrite(node.getValue(), context));
+        }
+    }
+
+    private enum NodeContext
+    {
+        ROOT_NODE,
+        NOT_ROOT_NODE;
+
+        boolean isRootNode()
+        {
+            return this == ROOT_NODE;
+        }
+    }
+
+    private static class ExtractCommonPredicatesExpressionRewriter
+            extends ExpressionRewriter<NodeContext>
+    {
+        @Override
+        public Expression rewriteExpression(Expression node, NodeContext context, ExpressionTreeRewriter<NodeContext> treeRewriter)
+        {
+            if (context.isRootNode()) {
+                return treeRewriter.rewrite(node, NodeContext.NOT_ROOT_NODE);
+            }
+
+            return null;
+        }
+
+        @Override
+        public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, NodeContext context, ExpressionTreeRewriter<NodeContext> treeRewriter)
+        {
+            List<Expression> predicates = extractPredicates(node.getType(), node).stream()
+                    .map(expression -> treeRewriter.rewrite(expression, NodeContext.NOT_ROOT_NODE))
+                    .collect(toList());
+
+            List<List<Expression>> subPredicates = getSubPredicates(predicates);
+            List<Expression> leafPredicates = getLeafPredicates(predicates);
```

Comment:
Indeed, there is no point in special leaves treatment, since simplification of `(A and B) or (A and C) or (A)` does not bring much value as well as selective simplification of predicates subset.